src/core/thread/multicast_subscription_table.hpp

```diff
@@ -0,0 +1,228 @@
+         * the constructor. If there are no more `MulticastSubscription` entries matching the given filter, the iterator
+         * becomes empty (i.e. `IsDone()` returns `true`).
+         *
+         */
+        void Advance(void);
+
+        /**
+         * This method advances the iterator.
+         *
+         * The iterator is moved to point to the next `MulticastSubscription` entry matching the given mask filter in
+         * the constructor. If there are no more `MulticastSubscription` entries matching the given filter, the iterator
+         * becomes empty (i.e. `IsDone()` returns `true`).
+         *
+         */
+        bool IsDone(void) { return mCurrent == nullptr; }
+
+        /**
+         * This method gets the `MulticastSubscription` entry to which the iterator is currently pointing.
+         *
+         * @returns A reference to the `MulticastSubscription` entry.
+         *
+         */
+        MulticastSubscription &GetMulticastSubscription(void) { return *mCurrent; }
+
+    private:
+        Filter                 mFilter;
+        MulticastSubscription *mCurrent;
+    };
+
+    /**
+     * This constructor initializes the Multicast Subscription Table object.
+     *
+     * @param[in]  aInstance     A reference to the OpenThread instance.
+     *
+     */
+    explicit MulticastSubscriptionTable(Instance &aInstance);
+
+    /**
+     * This method searches the Multicast Subscription Table for a `MulticastSubscription` with a given address also
+     * matching a given filter.
+     *
+     * @param[in] aAddress  The IP6 address.
+     *
+     * @returns A pointer to the `MulticastSubscription` entry if found, or `nullptr` otherwise.
+     *
+     */
+    MulticastSubscription *Find(const Ip6::Address &aAddress);
+    /**
+     * This method gets a new/unused `MulticastSubscription` entry from the Multicast Subscription Table.
+     *
+     * @note The returned listener entry will be cleared (`memset` to zero).
+     *
+     * @returns A pointer to a new `MulticastSubscription` entry, or `nullptr` if all `MulticastSubscription` entries
+     * are in use.
+     */
+    MulticastSubscription *New(void);
+
+    size_t Count(Filter aFilter);
+
+    void        SetLocallySubscribed(const Ip6::Address &aAddress, bool aLocallySubscribed);
+    static void SetLocallySubscribed(MulticastSubscription &aSubscription, bool aLocallySubscribed);
+
+    static void LogSubscription(const char *aAction, const Ip6::Address &aAddress, otError aError);
+
+    void SetRegistering(uint8_t aNum);
+    void ClearAllRegistered(void);
+    void UpdateRegisteringSubscriptions(bool aRegistered);
+
+    void Print(void);
+
+private:
+    enum
+    {
+        kMaxMulticastSubscriptionNum = OPENTHREAD_CONFIG_MAX_MULTICAST_SUBSCRIPTION_NUM,
+    };
+
+    static bool MatchesFilter(const MulticastSubscription &aSubscription, Filter aFilter);
+
+    MulticastSubscription mSubscriptions[kMaxMulticastSubscriptionNum];
```

Comment:
We believe it's not efficient to directly use MAs in `Child` and Netif. For example, if one child subscribes to `ff04::1`, the parent will have to traverse all addresses of all children (there might be hundreds of them) to determine if this is a newly subscribed address that needs a MLR.req. Same happens when a Child unsubscribes to a address.

To the opposite, we considered moving multicast addresses from Netif and Child to Multicast Table because `Multicast Table` is much more efficient in memory footprints if many Children are subscribing the same multicast address, which we expect to be a common case. 

I prefer to enhance Child in a separate PR.

Reply:
Thread currently requires maintaining at least 4 IPv6 addresses (unicast or multicast) per child. Are we proposing to size the subscription table to accommodate this worst case?

Reply:
That's a good point. Since ML-EID is a unicast address, we need to support at least 3 MA for each Child. 
 If a parent has maximum 10 Children, it will needs to have at least 30 proxied Multicast Subscriptions (otherwise it's a violation of Thread Spec.) .
Meanwhile, we can't reduce the size of Address list of the Child because we still need to support 4 Unicast Addresses. 

I think having to support `ChildrenNum * 3`  Multicast Subscriptions (19B per) can be a burden for routers. 

From such viewpoint, moving Multicast Addresses to this table can not bring us memory optimization. But the Child will be able to subscribe to a lot of Multicast Addresses. But subscribing to so many Multicast Addresses might not be a common use case for Thread networks. 

That being said, my opinion would be these:
- Using a separate table for proxying multicast addresses have significant performance advantage over using addresses in `Child`. 
- Storing a copy of multicast address on `Netif` to this table can make the implementation of `MlrManager` simpler, but it's less efficient in memory usage than using addersses in Netif directly. **But if the locally subscribed multicast address is also subscribed by a Child (proxied), there is no additional memory cost for copying it from Netif.**
- I expect the Thread nodes are subscribing to the same set of multicast addresses in a common setup. In such case, this table would contain addresses that are both locally subscribed and proxied, which is a very efficient use of memory.

Reply:
I think it is difficult to tell what the use of multicast groups will look like. In a home environment, I agree that the number of multicast groups may be small. In a commercial environment, the number of multicast groups may be much larger - Thread 1.2 Commercial is looking to handling hundreds of multicast groups, for example.

I agree that having a dedicated multicast address table for reports will be more efficient from a processing perspective. I am not so sure it saves on memory because we already allocate 3 addresses per child.

For unicast addresses, we have similar processing overhead:

1. When forwarding IPv6 packets, we need to scan potentially hundreds of IPv6 addresses to determine which child.

1. When performing duplicate address detection, we need to scan potentially hundreds of IPv6 addresses to ensure uniqueness across all children.

Reply:
I think it's apparent that using existing `NetifMulticastAddress` in `Netif` without copying them to a `MulticastListenerRegistrationTable` saves us RAM, especially for MTDs.
**I will create another PR based on this PR to refactor it so that no address is copied from `Netif` and not table is used.**
Let's then see the costs of each implementation and decide which to use. 

As for the proxied MLR.req, we analyzed the logic of sending MLR.req without using any table and found that the time complexity could be `O(ChildNum * IpNumPerChild * IP6AddressesTlvSize)`. On the other hand, the current implementation using a table has a time complexity of `O(MulticastIpNum)`. I think we can discuss more in the proxied MLR.req PR #5236 .

Thoughts? @librasungirl @jwhui @abtink

Reply:
Continuing the proxy discussion in https://github.com/openthread/openthread/pull/5236#discussion_r455500810